core/src/main/scala/org/apache/spark/deploy/history/HistoryAppStatusStore.scala

```diff
@@ -0,0 +1,123 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import scala.util.matching.Regex
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.History.{APPLY_CUSTOM_EXECUTOR_LOG_URL_TO_INCOMPLETE_APP, CUSTOM_EXECUTOR_LOG_URL}
+import org.apache.spark.status.{AppStatusListener, AppStatusStore}
+import org.apache.spark.status.api.v1
+import org.apache.spark.status.api.v1.ExecutorSummary
+import org.apache.spark.util.kvstore.KVStore
+
+private[spark] class HistoryAppStatusStore(
+    conf: SparkConf,
+    store: KVStore,
+    listener: Option[AppStatusListener] = None)
+  extends AppStatusStore(store, listener) with Logging {
+
+  import HistoryAppStatusStore._
+
+  private val logUrlPattern: Option[String] = conf.get(CUSTOM_EXECUTOR_LOG_URL)
+  private val applyReplaceLogUrlToIncompleteApp: Boolean =
+    conf.get(APPLY_CUSTOM_EXECUTOR_LOG_URL_TO_INCOMPLETE_APP)
+
+  override def executorList(activeOnly: Boolean): Seq[v1.ExecutorSummary] = {
+    val execList = super.executorList(activeOnly)
+    logUrlPattern match {
+      case Some(pattern) =>
+        if (applyReplaceLogUrlToIncompleteApp || isApplicationCompleted) {
+          execList.map(replaceLogUrls(_, pattern))
+        } else {
+          execList
+        }
+
+      case None => execList
+    }
+  }
+
+  override def executorSummary(executorId: String): v1.ExecutorSummary = {
+    val execSummary = super.executorSummary(executorId)
+    logUrlPattern match {
+      case Some(pattern) =>
+        if (applyReplaceLogUrlToIncompleteApp || isApplicationCompleted) {
+          replaceLogUrls(execSummary, pattern)
+        } else {
+          execSummary
+        }
+
+      case None => execSummary
+    }
+  }
+
+  private def isApplicationCompleted: Boolean = {
+    val appInfo = super.applicationInfo()
+    appInfo.attempts.nonEmpty && appInfo.attempts.head.completed
+  }
+
+  private def replaceLogUrls(exec: ExecutorSummary, urlPattern: String): ExecutorSummary = {
+    val attributes = exec.attributes
+
+    val allPatterns = CUSTOM_URL_PATTERN_REGEX.findAllMatchIn(urlPattern).map(_.group(1)).toSet
+    val allPatternsExceptFileName = allPatterns.filter(_ != "FILE_NAME")
+    val allAttributeKeys = attributes.keySet
+    val allAttributeKeysExceptLogFiles = allAttributeKeys.filter(_ != "LOG_FILES")
+
+    if (allPatternsExceptFileName.diff(allAttributeKeysExceptLogFiles).nonEmpty) {
+      logFailToRenewLogUrls("some of required attributes are missing in app's event log.",
```

Comment:
Are these warnings gonna show up every time an old event log is read (i.e. no extra attributes recorded)? That sounds pretty noisy and scary for users.

Reply:
Totally agreed. Thanks for finding! Maybe we have flag here and show only once per instance.
It would show up again when SparkUI for application is invalidated and recreated, but I don't think it will be noisy.

Reply:
I think it would be better to avoid a warning if the executor doesn't have any attributes recorded. That's either an old event log or from an application that doesn't support this feature.

If you just log an info message that is less noisy than this, it should be fine in that case. Leave the warning for when the app is actually missing info.

Reply:
OK. I made a change like I commented earlier, but agree to change the log level to INFO. Let me change it. Thanks!